pkg/yurttunnel/dns/dns.go

```diff
@@ -0,0 +1,463 @@
+	return dnsctl.updateTunnelServerSvcDnatPorts(dnatPorts)
+}
+
+func (dnsctl *coreDNSRecordController) syncDNSRecordAsWhole() {
+	klog.V(2).Info("sync dns record as whole")
+
+	dnsctl.lock.Lock()
+	defer dnsctl.lock.Unlock()
+
+	ip, err := dnsctl.getTunnelServerIP(false)
+	if err != nil {
+		klog.Errorf("failed to sync dns record as whole, %v", err)
+		return
+	}
+
+	nodes, err := dnsctl.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("failed to sync dns record as whole, %v", err)
+		return
+	}
+
+	records := make([]string, 0, len(nodes))
+	for i := range nodes {
+		node := nodes[i]
+		if !isEdgeNode(node) {
+			ip, err = getNodeHostIP(node)
+			if err != nil {
+				klog.Errorf("failed to parse node address for %v, %v", node.Name, err)
+				continue
+			}
+		}
+		records = append(records, formatDNSRecord(ip, node.Name))
+	}
+
+	if err := dnsctl.updateDNSRecords(records); err != nil {
+		klog.Errorf("failed to sync dns record as whole, %v", err)
+	}
+}
+
+func (dnsctl *coreDNSRecordController) getTunnelServerIP(useCache bool) (string, error) {
+	if useCache && len(dnsctl.tunnelServerIP) != 0 {
+		return dnsctl.tunnelServerIP, nil
+	}
+
+	svc, err := dnsctl.kubeClient.CoreV1().Services(constants.YurttunnelServerServiceNs).
+		Get(constants.YurttunnelServerServiceName, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get %v/%v service, %v",
+			constants.YurttunnelServerServiceNs, constants.YurttunnelServerServiceName, err)
+	}
+	if len(svc.Spec.ClusterIP) == 0 {
+		return "", fmt.Errorf("unable find ClusterIP from %s/%s service, %v",
+			constants.YurttunnelServerServiceNs, constants.YurttunnelServerServiceName, err)
+	}
+
+	// cache result
+	dnsctl.tunnelServerIP = svc.Spec.ClusterIP
+
+	return dnsctl.tunnelServerIP, nil
+}
+
+func (dnsctl *coreDNSRecordController) updateDNSRecords(records []string) error {
+	// keep sorted
+	sort.Strings(records)
+
+	cm, err := dnsctl.kubeClient.CoreV1().ConfigMaps(constants.YurttunnelServerServiceNs).
+		Get(yurttunnelDNSRecordConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	cm.Data[yurttunnelDNSRecordNodeDataKey] = strings.Join(records, "\n")
+	if _, err := dnsctl.kubeClient.CoreV1().ConfigMaps(constants.YurttunnelServerServiceNs).Update(cm); err != nil {
+		return fmt.Errorf("failed to update configmap %v/%v, %v",
+			constants.YurttunnelServerServiceNs, yurttunnelDNSRecordConfigMapName, err)
+	}
+	return nil
+}
+
+func (dnsctl *coreDNSRecordController) updateTunnelServerSvcDnatPorts(ports []string) error {
+	svc, err := dnsctl.kubeClient.CoreV1().Services(constants.YurttunnelServerServiceNs).
+		Get(constants.YurttunnelServerServiceName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to sync tunnel server service, %v", err)
+	}
+
+	// parse tunnel server tcp port
+	tunnelServerAgentPort, _ := strconv.Atoi(constants.YurttunnelServerAgentPort)
```

Comment:
tunnelServerAgentPort for insecure request is not 10262,  is 10264(YurttunnelServerMasterInsecurePort)